sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/Utils.java

```diff
@@ -0,0 +1,228 @@
+import com.azure.core.exception.ResourceNotFoundException;
+import com.azure.core.http.HttpHeaders;
+import com.azure.core.http.HttpResponse;
+import com.azure.core.http.rest.PagedResponse;
+import com.azure.core.http.rest.PagedResponseBase;
+import com.azure.core.http.rest.Response;
+import com.azure.core.http.rest.ResponseBase;
+import com.azure.core.util.CoreUtils;
+
+import java.util.List;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+final class Utils {
+    private static final String CONTINUATIONLINK_HEADER_NAME = "Link";
+    private static final String LINK_MATCHER = "(<(.+)>;.*)";
+    private static final Pattern CONTINUATIONLINK_PATTERN = Pattern.compile(LINK_MATCHER);
+
+    static <T> PagedResponse<T> getPagedResponseWithContinuationToken(PagedResponse<T> listResponse) {
+        return Utils.getPagedResponseWithContinuationToken(listResponse, values -> values);
+    }
+
+    static <T, R> PagedResponse<T> getPagedResponseWithContinuationToken(PagedResponse<R> listResponse, Function<List<R>, List<T>> mapperFunction) {
+        String continuationLink = null;
+        HttpHeaders headers = listResponse.getHeaders();
+
+        if (headers != null) {
+            String continuationLinkHeader = headers.getValue(CONTINUATIONLINK_HEADER_NAME);
+            if (!CoreUtils.isNullOrEmpty(continuationLinkHeader)) {
+                Matcher matcher = CONTINUATIONLINK_PATTERN.matcher(continuationLinkHeader);
+                if (matcher.matches()) {
+                    if (matcher.groupCount() == 2) {
+                        String apiPath = matcher.group(2);
+                        continuationLink = apiPath;
+                    }
+                }
+            }
+        }
+
+        List<T> values = null;
+
+        if (mapperFunction != null) {
+            values = mapperFunction.apply(listResponse.getValue());
+        }
+
+        return new PagedResponseBase<String, T>(
+            listResponse.getRequest(),
+            listResponse.getStatusCode(),
+            listResponse.getHeaders(),
+            values,
+            continuationLink,
+            null
+        );
+    }
+
+    static Throwable mapException(Throwable exception) {
+        if (!(exception instanceof AcrErrorsException)) {
+            return exception;
+        }
+
+        final AcrErrorsException errorsException = ((AcrErrorsException) exception);
+        final HttpResponse errorHttpResponse = errorsException.getResponse();
+
+        final int statusCode = errorHttpResponse.getStatusCode();
+        final String errorDetail = errorsException.getMessage();
+
+        switch (statusCode) {
+            case 401:
+                return new ClientAuthenticationException(errorDetail, errorsException.getResponse(), exception);
+            case 404:
+                return new ResourceNotFoundException(errorDetail, errorsException.getResponse(), exception);
```

Comment:
@jianghaolu, @srnagar could code generation be updated so that `401` and `404` use these sub-types of `HttpResponseException`? Should be non-breaking

Reply:
Yes, this could be either done through autorest which will require every service defining it or can be done in RestProxy if there's no exception mapping for these status codes.

Reply:
Created a work-item for this. - [https://github.com/Azure/azure-sdk-for-java/issues/20468](https://github.com/Azure/azure-sdk-for-java/issues/20468)

---
In reply to: [609653166](https://github.com/Azure/azure-sdk-for-java/pull/20368#discussion_r609653166) [](ancestors = 609653166)